### PR TITLE
CSE (Networks) & ESC (Finals Revision, Java Concurrency) notes

### PR DESCRIPTION
**Pull request by lyqht**

### CODE REVIEW

**Comment by lyqht** [comment]
Please let me know if my markdown files are too messy, and how i can improve them!

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 1]
```diff
@@ -0,0 +1,623 @@
+# 50.003 Finals Revision Document
+### *Credits:*
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: I think we can shift this to the end of the document.
I think "Acknowledgements" sound nicer than credits also.

Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: For the headings, adding `#` is enough. Don't add italics. Makes your document inconsistent with every other page.

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+- [LucidCharts](https://www.youtube.com/watch?v=pCK6prSq8aw) - UML Sequence Diagram
+- [Guru99](https://www.guru99.com/equivalence-partitioning-boundary-value-analysis.html) - Equivalence Class Partitioning & Boundary Value Testing
+- [Tester Stories](http://testerstories.com/2014/06/path-testing-the-coverage/) - Control Flow Graphs
+- [StackOverflow](https://stackoverflow.com/questions/22351222/can-we-change-the-value-of-a-final-variable-of-a-mutable-class) - OBJ04-J's explanation on the`final` term 
+- 50.003 Elements of Software Construction Slides - Syllabus and some content
+- [Prof Sudipta](https://istd.sutd.edu.sg/people/faculty/sudipta-chattopadhyay) - Comments on Finals
+- [Lyqht](https://github.com/lyqht) - Compilation
+
+
+### *Table of Contents*
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: There's a lot of content, which is good. However, can consider splitting into multiple pages.

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+      - [Equivalence Class Partitioning & Boundary Value Testing](#equivalence-class-partitioning---boundary-value-testing)
+    + [Fault-based Testing based on Dirty / Failure Test Cases](#fault-based-testing-based-on-dirty---failure-test-cases)
+    + [All coverage metrics](#all-coverage-metrics)
+      - [Control Flow Graph](#control-flow-graph)
+  * [Test Generation](#test-generation)
+    + [Instrumentation](#instrumentation)
+    + [Genetic Algorithm](#genetic-algorithm)
+    + [Symbolic Execution](#symbolic-execution)
+- [Concurrency](#concurrency)
+  * [The Need for Refactoring](#the-need-for-refactoring)
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Refactoring is under concurrency?

The content below seems to all be under the concurrency heading.

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+    - Making statements atomic alone does not solve the issue of atomicity. Often according to your requirement, you need to make multiple statements (a code block) atomic. 
+    - e.g. consider a stack class that is trying to pop an item. You might need to check the empty stack and the pop together atomically. If only single statements are atomic, this only guarantees that the check and the updates are atomic in isolation, but not together. 
+    - To solve such problem, we will still need locks. 
+3. **Execution Ordering**
+    - Unless the hardware implements a completely deterministic scheduler known to the programmer, the execution order problem will persist. In practice, it is almost impossible to implement a completely deterministic scheduler. 
+    - But, using locks can help to restrict the number of possible execution orders. 
+
+*refer to [Java Concurrency notes](https://hackmd.io/iynzv5vhSRyUL9gusM71CQ) for more info on **implementation**.*
+
+## The Need for Refactoring
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Mistake in heading level?

* `#` top level is concurrency
* `##` is refactoring

Is this a mistake?

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+    - A description of a set of sequences of actions, including variants, that system performs that yields an observable value to an actor. 
+    - Should ideally begin with a verb.
+    - Think of the end goal of a user. They don't want to "login" or "sign up." That's not a use case. The use case is more like "make a purchase."
+
+- Actors
+    - People or systems that provide or receive information from the system; they are among the stakeholders of a system, which could be human beings, other systems, timers and clocks or hardware devices.
+    - Primary Actors: stimulate the system and are the initiators of events (active).
+    - Secondary Actors: only receive stimuli from the system (passive).
+
+// ***example***: of use case diagram
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: What's with the `//`?

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+- **Link Coverage:** Achieved when the paths identified have a test that goes along every link, or line, in the graph. In many cases, node coverage will take care of this.
+- **Loop Coverage:** Achieved when the numerous paths identified have tests that explores the interaction between sub-paths within a loop.
+- **Statement coverage**: Achieved when the flow of control reached every executable statement of source code at least once.
+- **Path Coverage**: Achieved when all linearly independent paths in the program are *executed at least once*
+    - A path is defined as the number of nodes executed.
+    - A linearly independent path can be defined in terms of what’s called a *control flow graph* of an application.
+    - If there is a while loop happening, depending on the input, there can be a lot of paths such that it is impossible to achieve path coverage.
+
+// ***example***: code that has a loop and impossible to achieve path coverage.
+```
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Add language marker to code block. E.g:
```
```java
codecodecode
`` `
``` 
gives you syntax highlighting in the code:
```java
int f1(int x) {
    int y;
    while (x < 0) {
       if (x >= 0) 
            y = 1/x;
        else
            y++;
        x ++;
    }
}
```

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+    - Application programmer interfaces (APIs)
+    - Choices of data structure and algorithms
+
+- Things that can be included:
+    - General description of what it does ***ONLY IF** your code cannot be self-explanatory.* e.g. your variables and method names by right should already speak for themselves what they do. Unless coz of legacy code that you can't override and have to use them.
+    - Precondition (@param), Postcondition (output)
+
+
+```
+//precondition: true
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Add space after `//` looks better.

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+
+## The Need for Refactoring
+### Comments
+- Used to document
+    - Application programmer interfaces (APIs)
+    - Choices of data structure and algorithms
+
+- Things that can be included:
+    - General description of what it does ***ONLY IF** your code cannot be self-explanatory.* e.g. your variables and method names by right should already speak for themselves what they do. Unless coz of legacy code that you can't override and have to use them.
+    - Precondition (@param), Postcondition (output)
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Wrap the code segments like this `@param`

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 9]
```diff
@@ -0,0 +1,623 @@
...
+*refer to [Java Concurrency notes](https://hackmd.io/iynzv5vhSRyUL9gusM71CQ) for more info on **implementation**.*
+
+## The Need for Refactoring
+### Comments
+- Used to document
+    - Application programmer interfaces (APIs)
+    - Choices of data structure and algorithms
+
+- Things that can be included:
+    - General description of what it does ***ONLY IF** your code cannot be self-explanatory.* e.g. your variables and method names by right should already speak for themselves what they do. Unless coz of legacy code that you can't override and have to use them.
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Don't use the "coz" as it is too informal

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+// Found black listed tag
+throw new IllegalStateException();
+} else {
+// . . .
+}
+```
+
+Also, Non-valid ASCII characters should be removed first before normalizing the code, if not these characters may slip unnoticed through the normalizer and be run.
+
+So this meant:
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: So this means? Grammar

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+    outStream.write(xmlString.getBytes());
+    outStream.flush();
+}
+```
+
+### SQL Injection
+An SQL injection vulnerability arises when the original SQL query can be altered to form a different query.
+
+> An SQL command to authenticate a user might take the form:
+`SELECT * FROM db_user WHERE username='<USERNAME>' AND
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Even though this is a single line, wrap in a multi line code block (triple `) and add language annotation for syntax highlighting:
```sql
SELECT * FROM db_user WHERE username='<USERNAME>' AND password='<PASSWORD>'
```

File: 50.003 Elements of Software Construction/Java Concurrency.md [hunk 12]
```diff
@@ -0,0 +1,473 @@
+# Java Concurrency
+### *Credits:*
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/Java Concurrency.md]: Same thing, move credits to back

File: 50.003 Elements of Software Construction/Java Concurrency.md [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+An algorithm is called non-blocking if failure or suspension of any thread cannot cause failure or suspension of another thread. 
+
+Non-blocking algorithms are immune to deadlock (though, in unlikely scenarios, may exhibit livelock or starvation)
+
+## Java Synchronizers
+### CountdownLatch
+The CountDownLatch class allows one or more threads to wait until a set of operations in other threads complete. CountDownLatch works by having a counter initialized with number of threads, which is decremented each time a thread complete its execution. When count reaches to zero, it means all threads have completed their execution, and thread waiting on latch resume the execution.
+
+1. `CountDownLatch latch = new CountDownLatch(numberOfThreads);
+`
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/Java Concurrency.md]: Added new line by accident?

File: 50.003 Elements of Software Construction/Java Concurrency.md [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+        }
+    }
+    
+    public doSomeWork() {
+        // some working code
+    }
+    
+```
+
+What if you are not certain of how many batches of tasks you actually need to complete? You would need to create a new CountDownLatch for every batch of tasks that you receive, and in this code, this is not accounted for. So, we turn to CyclicBarrier for resolving this issue.
```
Comment (tlkh) [inline 50.003 Elements of Software Construction/Java Concurrency.md]: Wording can be more precise. Maybe something like:

Following the method shown by this code, a new `CountDownLatch` object would need to somehow be created for every batch of tasks if you are uncertain of the total number of batches. To address this problem in an elegant manner, we can use `CyclicBarrier` (next section)

File: 50.005 Computer System Engineering/2018/Networking.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+- Usually, the mode of transmission of packets is:
+`PC -> Switch -> Router -> Modem -> Modem -> Switch -> Servers`
+
+- When the packets arrive at the router, the packet arrival rate to link temporarily exceeds the output link capacity. Hence they would queue in the router buffers, causing a delay.
+    - If there are no free buffers for them to get into queue, then the packets will be dropped. This is known as packet loss.
+
+### Single Router Packet Delay Sources:
+
+![](https://i.imgur.com/V2iaWr9.png)
```
Comment (tlkh) [inline 50.005 Computer System Engineering/2018/Networking.md]: For all the diagrams, make sure you have Natalie's blessing to submit here.

Even so, I would like for us to eventually replace these with our own diagrams.

File: 50.005 Computer System Engineering/2018/Networking.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    - Waiting time for packet to get to front of queue for output link
+    - Depends on congestion level (how much other users are also sending data)
+3. Transmission *(dtrans)*
+    - Time to push the whole packet *(all the bits)* from router to link 
+    - *dtrans* = packet length (bits) / link bandwith (bps)
+`dtrans = L/R` 
+4. Propagation *(dprop)* 
+    - Time for packet to move from beginning to end of the link
+    - *dprop* = length of physical link (m) / propagation speed in medium *(~2x10^8 m/s - 2/3 speed of light in vacuum)*
+`dprop = d/s` **m/s**
```
Comment (tlkh) [inline 50.005 Computer System Engineering/2018/Networking.md]: The uh, formula and units be clearer. Now it looks awkward. Suggest placing everything inside code segment.

For formulas, you can try exporting HTML from https://www.codecogs.com/latex/eqneditor.php as well

<a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;\verb|dprop|&space;=&space;\frac{d}{s}&space;\mathrm{m/s}" target="_blank"></a>

File: 50.005 Computer System Engineering/2018/Networking.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+### Modems 
+- Carries out Modulation & Demodulation
+- Modulation: A process to add text/img/vid/sound etc into a carrier signal. Demodulation is vice versa.
+    - Types of broadbands:
+        - Dial-up: audio freq <-> digital
+        - DSL: copper telephone lines
+        - Coaxial Cables: faster than DSL
+        - Fibre Optic Cables: Light pulses
+    - Wifi signals: EM waves
```
Comment (tlkh) [inline 50.005 Computer System Engineering/2018/Networking.md:104]: Wi-Fi

**Review by tlkh — COMMENTED** [review]
Good contribution!

**Comment by tlkh** [comment]
@lyqht please look at my comments and make changes accordingly. No need to rush, do it properly and then just **push over your current fork**. You can edit your current commit too, by doing:
```bash
git commit --amend
git push
```

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md (hunk 1 above)
Comment (lyqht) [reply 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: Okay! I'll add this

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md (hunk 2 above)
Comment (lyqht) [reply 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: do you mean multiple documents?

File: 50.003 Elements of Software Construction/Java Concurrency.md (hunk 13 above)
Comment (lyqht) [reply 50.003 Elements of Software Construction/Java Concurrency.md]: Yes its a mistake haha

File: 50.003 Elements of Software Construction/50.003 Finals Revision Document.md (hunk 1 above)
Comment (lyqht) [reply 50.003 Elements of Software Construction/50.003 Finals Revision Document.md]: okay

File: 50.003 Elements of Software Construction/Java Concurrency.md (hunk 14 above)
Comment (lyqht) [reply 50.003 Elements of Software Construction/Java Concurrency.md]: Great advice, thanks!

File: 50.005 Computer System Engineering/2018/Networking.md (hunk 15 above)
Comment (lyqht) [reply 50.005 Computer System Engineering/2018/Networking.md]: Okay!

**Comment by lyqht** [comment]
Thanks for the in-depth feedback! I'm not familiar with how certain things should be formatted so pardon the messiness. I'm not sure how to emphasize the example sections so i just added a // in front, what is a better way of doing so?

**Comment by joel-huang** [comment]
@tlkh @lyqht @flamanta Just a reminder if this got forgotten - do you guys want to merge this in first?

**Comment by lyqht** [comment]
hi sry for the delay! i've improved the formatting after @tlkh 's suggestions. removed most of the images that were drawn by our prof.

**Comment by flamanta** [comment]
Looks fine to me! If you don't mind, I will convert the Markdown files into a PDF file so that the LaTeX would render properly. 

Could you merge the changes from `upstream/master` into your local `master` branch?

**Comment by tlkh** [comment]
Thanks guys
